drivers/input/input_sbus.c

```diff
@@ -0,0 +1,363 @@
+			      unsigned int value)
+{
+	const struct input_sbus_config *const config = dev->config;
+	struct input_sbus_data *const data = dev->data;
+
+	int channel = data->channel_mapping[sbus_channel];
+
+	/* Not Mapped */
+	if (channel == -1) {
+		return;
+	}
+
+	if (value > (data->last_reported_value[channel] + REPORT_FILTER) ||
+	    value < (data->last_reported_value[channel] - REPORT_FILTER)) {
+		switch (config->channel_info[channel].type) {
+		case INPUT_EV_ABS:
+		case INPUT_EV_MSC:
+			input_report(dev, config->channel_info[channel].type,
+				     config->channel_info[channel].zephyr_code, value, false,
+				     K_FOREVER);
+			break;
+
+		default:
+			if (value > CHANNEL_VALUE_ONE) {
+				input_report_key(dev, config->channel_info[channel].zephyr_code, 1,
+						 false, K_FOREVER);
+			} else if (value < CHANNEL_VALUE_ZERO) {
+				input_report_key(dev, config->channel_info[channel].zephyr_code, 0,
+						 false, K_FOREVER);
+			}
+		}
+		data->last_reported_value[channel] = value;
+	}
+}
+
+static void input_sbus_input_report_thread(const struct device *dev, void *dummy2, void *dummy3)
+{
+	struct input_sbus_data *const data = dev->data;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	uint8_t i, channel;
+	uint8_t *sbus_channel_data = &data->sbus_frame[1]; /* Omit header */
+	uint16_t value;
+	int bits_read;
+	unsigned int key;
+	int ret;
+	bool connected_reported = false;
+
+	while (true) {
+		if (!data->in_sync) {
+			k_sem_take(&data->report_lock, K_FOREVER);
+			if (data->in_sync) {
+				LOG_DBG("SBUS receiver connected");
+			} else {
+				continue;
+			}
+		} else {
+			ret = k_sem_take(&data->report_lock, K_MSEC(SBUS_INTERFRAME_SPACING));
+			if (ret == -EBUSY) {
+				continue;
+			} else if (ret < 0 || !data->in_sync) {
+				/* We've lost sync with the UART receiver */
+				key = irq_lock();
+
+				data->partial_sync = false;
+				data->in_sync = false;
+				data->xfer_bytes = 0;
+				irq_unlock(key);
+
+				connected_reported = false;
+				LOG_DBG("SBUS receiver connection lost");
+
+				/* Report connection lost */
+				continue;
+			}
+		}
+
+		if (connected_reported &&
+		    data->sbus_frame[SBUS_BYTE24_IDX] & SBUS_BYTE24_FRAME_LOST) {
+			LOG_DBG("SBUS controller connection lost");
+			connected_reported = false;
+		} else if (!connected_reported &&
+			   !(data->sbus_frame[SBUS_BYTE24_IDX] & SBUS_BYTE24_FRAME_LOST)) {
+			LOG_DBG("SBUS controller connected");
+			connected_reported = true;
+		}
+
+		/* Parse the data */
+		channel = 0;
+		value = 0;
+		bits_read = 0;
+
+		for (i = 0; i < SBUS_SERVO_LEN; i++) {
+			/* Read the next byte */
+			unsigned char byte = sbus_channel_data[i];
+
+			/* Extract bits and construct the 11-bit value */
+			value |= byte << bits_read;
+			bits_read += 8;
+
+			/* Check if we've read enough bits to form a full 11-bit value */
+			while (bits_read >= 11) {
+				input_sbus_report(dev, channel, value & SBUS_SERVO_CH_MASK);
```

Comment:
the input logic would be to set the sync bit on the last event, maybe you can pass a sync bool here when `i == SBUS_SERVO_LEN - 1` and use it in the report?

Reply:
Oh wasn't aware of the `input_sync();` functionality, sure I can implement that.

Reply:
I see, yeah Linux implements a dedicated event for the sync message, we went for a bit in the normal messages to save a call. Maybe you can send an extra `input_report(dev, 0, 0, 0, true, K_FOREVER)` just so the application has something to work on.

Reply:
I doubt it's useful for everyone so I put it behind the `INPUT_SBUS_SEND_SYNC` symbol.

Reply:
I think it should be sent unconditionally, of at least have the option default to yes. Any generic code that relies on sync to process the data would stall forever.

Reply:
Made the default `y`.